timelock-agent/src/main/java/com/palantir/timelock/paxos/BatchTimelockPaxosLearnerAdapter.java

```diff
@@ -0,0 +1,108 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.timelock.paxos;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.Set;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSetMultimap;
+import com.google.common.collect.Iterables;
+import com.palantir.atlasdb.timelock.paxos.BatchPaxosLearnerRpcClient;
+import com.palantir.atlasdb.timelock.paxos.Client;
+import com.palantir.atlasdb.timelock.paxos.PaxosUseCase;
+import com.palantir.atlasdb.timelock.paxos.WithSeq;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.paxos.PaxosLearner;
+import com.palantir.paxos.PaxosValue;
+
+public class BatchTimelockPaxosLearnerAdapter implements PaxosLearner {
+    private final PaxosUseCase paxosUseCase;
+    private final Client client;
+    private final BatchPaxosLearnerRpcClient rpcClient;
+
+    private volatile long lastKnownSequence = Long.MIN_VALUE;
+
+    public BatchTimelockPaxosLearnerAdapter(
+            PaxosUseCase paxosUseCase,
+            Client client,
+            BatchPaxosLearnerRpcClient rpcClient) {
+        this.paxosUseCase = paxosUseCase;
+        this.client = client;
+        this.rpcClient = rpcClient;
+    }
+
+    public static PaxosLearner singleLeader(BatchPaxosLearnerRpcClient rpcClient) {
+        return new BatchTimelockPaxosLearnerAdapter(
+                PaxosUseCase.LEADER_FOR_ALL_CLIENTS,
+                PaxosUseCase.PSEUDO_LEADERSHIP_CLIENT,
+                rpcClient);
+    }
+
+    @Override
+    public void learn(long seq, PaxosValue val) {
+        rpcClient.learn(paxosUseCase, ImmutableSetMultimap.of(client, val));
+        updateLastKnownSequence(seq);
+    }
+
+    @Override
+    public Optional<PaxosValue> getLearnedValue(long seq) {
+        Set<PaxosValue> result = rpcClient.getLearnedValues(paxosUseCase, ImmutableSet.of(WithSeq.of(client, seq)))
+                .get(client);
+        checkResult(result);
+        if (result.isEmpty()) {
+            return Optional.empty();
+        }
+        updateLastKnownSequence(seq);
+        return Optional.of(Iterables.getOnlyElement(result));
+    }
+
+    @Override
+    public Optional<PaxosValue> getGreatestLearnedValue() {
+        Set<PaxosValue> result = rpcClient
+                .getLearnedValuesSince(paxosUseCase, ImmutableMap.of(client, lastKnownSequence))
+                .get(client);
+        Optional<PaxosValue> greatestIfExists = result.stream()
+                .reduce((fst, snd) -> fst.getRound() > snd.getRound() ? fst : snd);
```

Comment:
Equivalent, but I think I like the readability of this better:

```suggestion
                .max(Comparator.comparingLong(PaxosValue::getRound));
```